pylint/checkers/base_checker.py

```diff
@@ -0,0 +1,114 @@
+
+from typing import Any
+
+from pylint.config import OptionsProviderMixIn
+from pylint.exceptions import InvalidMessageError
+from pylint.interfaces import UNDEFINED
+from pylint.message import build_message_definition
+
+
+class BaseChecker(OptionsProviderMixIn):
+
+    # checker name (you may reuse an existing one)
+    name = None  # type: str
+    # options level (0 will be displaying in --help, 1 in --long-help)
+    level = 1
+    # ordered list of options to control the ckecker behaviour
+    options = ()  # type: Any
+    # messages issued by this checker
+    msgs = {}  # type: Any
+    # reports issued by this checker
+    reports = ()  # type: Any
+    # mark this checker as enabled or not.
+    enabled = True
+
+    def __init__(self, linter=None):
+        """checker instances should have the linter as argument
+
+        :param ILinter linter: is an object implementing ILinter."""
+        if self.name is not None:
+            self.name = self.name.lower()
+        OptionsProviderMixIn.__init__(self)
+        self.linter = linter
+
+    def __gt__(self, other):
+        """Permit to sort a list of Checker by name."""
+        return "{}{}".format(self.name, self.msgs).__gt__(
+            "{}{}".format(other.name, other.msgs)
+        )
+
+    def __repr__(self):
+        status = "Checker" if self.enabled else "Disabled checker"
+        msgids = [id for id in self.msgs]
+        return "{} '{}' responsible for {}".format(status, self.name, ", ".join(msgids))
+
+    def add_message(
+        self,
+        msgid,
+        line=None,
+        node=None,
+        args=None,
+        confidence=UNDEFINED,
+        col_offset=None,
+    ):
+        self.linter.add_message(msgid, line, node, args, confidence, col_offset)
+
+    def check_consistency(self) -> None:
+        """Check the consistency of msgid.
+
+        msg ids for a checker should be a string of len 4, where the two first
+        characters are the checker id and the two last the msg id in this
+        checker.
+
+        :raises InvalidMessageError: If the checker id in the messages are not
+        always the same. """
+        checker_id = None
+        existing_ids = []
+        for message in self.messages:
+            if checker_id is not None and checker_id != message.msgid[1:3]:
+                error_msg = "Inconsistent checker part in message id "
+                error_msg += "'{}' (expected 'x{checker_id}xx' ".format(
+                    message.msgid, checker_id=checker_id
+                )
+                error_msg += "because we already had {existing_ids}).".format(
+                    existing_ids=existing_ids
+                )
+                raise InvalidMessageError(error_msg)
+            checker_id = message.msgid[1:3]
+            existing_ids.append(message.msgid)
+
+    @property
+    def messages(self) -> list:
+        messages = []
+        for msgid, msg_tuple in sorted(self.msgs.items()):
```

Comment:
Can we squash this into a single list comprehension?
```suggestion
        return [build_message_definition(self, msgid, msg_tuple) for msgid, msg_tuple in sorted(self.msgs.items())]
```